roaring/src/bitmap/iter.rs

```diff
@@ -294,7 +316,337 @@ impl RoaringBitmap {
                 count += 1;
             }
         }
-
         Ok(count)
     }
 }
+
+mod iter_inner {
+    use crate::bitmap::container::Container;
+    use crate::bitmap::{container, util, IntoIter, Iter};
+    use core::slice;
+
+    pub(super) trait IterInternal<'a> {
+        type Container: IntoIterator<IntoIter = container::Iter<'a>, Item = u32> + AsRef<Container>;
+
+        type ContainerIterator: Iterator<Item = Self::Container> + DoubleEndedIterator;
+        type IntoContainerIterator: IntoIterator<IntoIter = Self::ContainerIterator>;
+
+        fn pop_container_front(&mut self) -> Option<Self::Container>;
+        fn pop_container_back(&mut self) -> Option<Self::Container>;
+
+        fn drain_containers_until(&mut self, index: usize);
+
+        fn clear_containers(&mut self);
+
+        fn find_container(&self, key: u16) -> Option<usize>;
+
+        fn iter_front_mut(&mut self) -> &mut container::Iter<'a>;
+        fn iter_back_mut(&mut self) -> &mut Option<container::Iter<'a>>;
```

Comment:
We should be able to avoid these, if we use `slice::Iter<'a, Container>` and `vec::IntoIter<Container>`, rather than `&[Container]` and `VecDeque<Container>`. Both implement `AsRef<[Container]>`, `Iterator` (for `next` + `nth`), and `Iterator::Item: IntoIterator<IntoIter=container::Iter>`, so it should be possible to write a function something like:

```rust
fn advance_to_impl<'a, It>(
    n: u32,
    front_iter: &mut Option<container::Iter<'a>>,
    containers: &mut It,
    back_iter: &mut Option<container::Iter<'a>>,
) where
    It: Iterator,
    It: AsRef<[Container]>,
    It::Item: IntoIterator<IntoIter = container::Iter<'a>>,
{
  let (key, index) = util::split(n);
  // Compare the key of n to the front iter's key {LT => Do nothing, EQ => advance front to index, GT => consume front iter, continue}
  
  // Binary search for the key in containers, if found, front_iter = containers.nth(idx) and advance it, else use _ = nth(idx - 1) to consume up to where it would be.

  // If containers is now empty, check the back iter's key, and possibly advance it or consume it
}
```

Which has all the logic that could be called by both advance_to implementations.